### ISSUE TITLE
Naming of aggregation feature which outputs multiple columns returns TypeError in feature_base/hash

### ISSUE THREAD

**Issue by johnnyheineken** [Featuretools/featuretools#591]
### Bug/Feature Request Description

This is similar to #540, as I have not found documentation for these kind of primitives.

I would like to create aggregation primitive, which creates multiple columns at once. However, naming is important part of this procedure, and I have achieved to add the clumns, but not to name them.

This is caused by `feature_base\feature_base.py in hash(self)`

This is my test benchmark:

```
import featuretools as ft
import pandas as pd
import numpy as np


CURRENT_FUNCTIONS = dict(
    max=np.nanmax,
    min=np.nanmin,
    mean=np.nanmean,
    sum=np.nansum,
    count=len,
    mean_monthly=lambda x: np.nanmean(x) / 12,
)


def agg_last_x_days(values, time_col, days, time=None):
    """
    Aggregate given data using prespecified aggregation functions.
    Possibilities are set in CURRENT_FUNCTIONS
    
    """
    data = values[
        time_col.values >= np.datetime64(time - np.timedelta64(days, "D"))
    ].values

    # only floats can have infinities
    if data.dtype.kind == "f":
        # we need to remove infinities before aggregation to be consistent with oracle
        data[np.isinf(data)] = np.nan
    result = []
    try:
        for function in CURRENT_FUNCTIONS.values():
            result += [function(data)]
#     except ValueError:
#         result = [None] * len(CURRENT_FUNCTIONS)
    except KeyError:
        raise KeyError("Unidentified aggregation function as parameter 'func'")
#     breakpoint()
    return result


def agg_last_x_days_generate_name(self, child_entity_id, base_feature_names, **kwargs):
    # print(self.kwargs)]})"
    names = ["{func}_{days}D({child_entity_id}.{feature}{where_string})".format(
            func=func.upper(),
            days=str(self.kwargs["days"]),
            child_entity_id=child_entity_id,
            feature=base_feature_names[0],
            where_string=kwargs["where_str"],
        ) for func in CURRENT_FUNCTIONS.keys()]
    return names


AGG_LAST_X_DAYS = ft.primitives.make_agg_primitive(
    function=agg_last_x_days,  # function to be used
    input_types=[
        ft.variable_types.Numeric,
        ft.variable_types.DatetimeTimeIndex,
    ],  # input data types
    number_output_features=len(CURRENT_FUNCTIONS),
    return_type=ft.variable_types.Numeric,  # data types to be returned
    uses_calc_time=True,  # whether function can utilize cutoff time information
    stack_on_self=False,  # whether the primitive could be stacked on itself
    cls_attributes={
        "generate_name": agg_last_x_days_generate_name
    },  # passing of name generating function
)

df = pd.DataFrame(
    {
        "ID": [1, 1, 1, 1, 2, 2],
        "dtime": [
            "2016-06-28 09:51:51.0",
            "2017-06-03 12:16:08.0",
            "2017-07-02 12:15:08.0",
            "2017-07-04 12:15:08.0",
            "2017-07-02 12:15:08.0",
            "2017-07-05 13:15:08.0",
        ],
        "value": [0, 3.0, 5, 0, 11, 14],  # float
        "value2": [1, 11, 1, 1, 11, 11],  # int
    }
)

cutoff_times = pd.DataFrame(
    {"ID": [1, 2], "time": ["2017-07-03 12:15:08.0", "2017-07-03 12:15:08.0"]}
)
df["dtime"] = pd.to_datetime(df["dtime"])
cutoff_times["time"] = pd.to_datetime(cutoff_times["time"])

es = ft.EntitySet("Dataset")
es.entity_from_dataframe(
    dataframe=df, entity_id="CAPP", index="index", time_index="dtime"
)

es.normalize_entity(
    base_entity_id="CAPP", new_entity_id="CLIENT", index="ID", make_time_index=False
)

fm, features = ft.dfs(
    entityset=es,
    target_entity="CLIENT",
    agg_primitives=[AGG_LAST_X_DAYS(days=1000)],
    features_only=False)

fm
```

this exact code raises:

```

2019-06-12 10:11:32,461 featuretools.entityset - WARNING    index index not found in dataframe, creating new integer column

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-68-ad74bb4230d5> in <module>
    102     target_entity="CLIENT",
    103     agg_primitives=[AGG_LAST_X_DAYS(days=1000)],
--> 104     features_only=False)
    105 
    106 fm

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\utils\entry_point.py in function_wrapper(*args, **kwargs)
     44                     ep.on_error(error=e,
     45                                 runtime=runtime)
---> 46                 raise e
     47 
     48             # send return value

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\utils\entry_point.py in function_wrapper(*args, **kwargs)
     36                 # call function
     37                 start = time.time()
---> 38                 return_value = func(*args, **kwargs)
     39                 runtime = time.time() - start
     40             except Exception as e:

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\synthesis\dfs.py in dfs(entities, relationships, entityset, target_entity, cutoff_time, instance_ids, agg_primitives, trans_primitives, groupby_trans_primitives, allowed_paths, max_depth, ignore_entities, ignore_variables, seed_features, drop_contains, drop_exact, where_primitives, max_features, cutoff_time_in_index, save_progress, features_only, training_window, approximate, chunk_size, n_jobs, dask_kwargs, verbose, return_variable_types)
    197 
    198     features = dfs_object.build_features(
--> 199         verbose=verbose, return_variable_types=return_variable_types)
    200 
    201     if features_only:

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\synthesis\deep_feature_synthesis.py in build_features(self, return_variable_types, verbose)
    225 
    226         self._run_dfs(self.es[self.target_entity_id], [],
--> 227                       all_features, max_depth=self.max_depth)
    228 
    229         new_features = list(all_features[self.target_entity_id].values())

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\synthesis\deep_feature_synthesis.py in _run_dfs(self, entity, entity_path, all_features, max_depth)
    334                                      child_entity=self.es[b_entity_id],
    335                                      all_features=all_features,
--> 336                                      max_depth=max_depth)
    337 
    338         """

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\synthesis\deep_feature_synthesis.py in _build_agg_features(self, all_features, parent_entity, child_entity, max_depth)
    595                                            parent_entity=parent_entity,
    596                                            primitive=agg_prim)
--> 597                 self._handle_new_feature(new_f, all_features)
    598 
    599                 # Obey allow where

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\synthesis\deep_feature_synthesis.py in _handle_new_feature(self, new_feature, all_features)
    410             return
    411         entity_id = new_feature.entity.id
--> 412         if new_feature.hash() in all_features[entity_id]:
    413             return
    414             raise Exception("DFS runtime error: tried to add feature %s"

c:\program files (x86)\microsoft visual studio\shared\python37_64\lib\site-packages\featuretools\feature_base\feature_base.py in hash(self)
    160 
    161     def hash(self):
--> 162         return hash(self.get_name() + self.entity.id)
    163 
    164     def __hash__(self):

TypeError: can only concatenate list (not "str") to list

```

This is caused by passing list of names into `cls.generate_name`, and this is killed inside hash function.

Unlike transformation primitive, aggregation primitive does not have `get_feature_names` function as mentioned in #540 

I am running this on Win10, using FT 0.8.0 & Python 3.7

**Comment by kmax12** [comment]
Currently, we do not support giving specific names to each of the output dimensions. When you set `number_output_features` featuretools will automatically append `__0` ... `__n-1` to your output features. 

If you remove your code for generate name, what you've done about should work to produce features.

If you want specific names, I'd recommend breaking your custom primitive into multiple functions for now.

**Comment by kmax12** [comment]
Opened issue #593 for this, so closing this issue for now. 

Feel free to follow up here if the proposal above doesn't work for you

**Comment by apemann** [comment]
Hi @kmax12. I've run into this same issue, and it doesn't seem to be fixed by pull #679.  

You've created a variable in FeatureBase to store a list of column names, but I can't see how you intend to set this value manually. 

The get_names function simply appends an index to the string passed by the custom generate_name function. This means it still isn't possible to give each column of a multi-output a custom name.

@rwedge proposed adding a generate_names function which returns a list, which could be used by the get_names function. Then you could simply do this:

```    
    def get_names(self):
        if not self._names:
            self._names = self.generate_names()
        return self._names
```

**Comment by rwedge** [comment]
@apemann ,

Thanks for binging this up,  that PR did not fix this issue!  I'll reopen #593.  I think the generate_names function is a good solution.
